File: src/bytecode/gvm.ts
import type { Chunk } from "./compiler.ts";
import { Op } from "./bytecode.ts";
import type { Literal } from "../token.ts";

/**GosVM */
export class GVM {
  private ip = 0;
  private stack: Literal[] = [];
  private slots: Literal[] = [];
  private callStack: { returnIp: number; baseSlot: number }[] = [];
  private currBaseSlot = 0;

  constructor(private chunk: Chunk, private maxSlot: number) {
    this.slots = new Array<Literal>(this.maxSlot);
  }

  public run = (): void => {
    const jmpTable: (() => void)[] = [];
    jmpTable[Op.LOAD_CONST] = () => {
      const idx = this.chunk.code[this.ip++];
      this.stack.push(this.chunk.constants[idx]);
    };
    jmpTable[Op.LOAD_VAR] = () => {
      const slot = this.chunk.code[this.ip++];
      this.stack.push(this.slots[this.currBaseSlot + slot]);
    };
    jmpTable[Op.STORE_VAR] = () => {
      const slot = this.chunk.code[this.ip++];
      this.slots[this.currBaseSlot + slot] = this.stack.pop();
    };
    jmpTable[Op.ADD] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = (left as any) + (right as any);
      this.stack.push(val);
    };
    jmpTable[Op.SUB] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = (left as number) - (right as number);
      this.stack.push(val);
    };
    jmpTable[Op.MUL] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = (left as number) * (right as number);
      this.stack.push(val);
    };
    jmpTable[Op.DIV] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = (left as number) / (right as number);
      this.stack.push(val);
    };
    jmpTable[Op.EQ] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = left === right;
      this.stack.push(val);
    };
    jmpTable[Op.NE] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = left !== right;
      this.stack.push(val);
    };
    jmpTable[Op.GT] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = left! > right!;
      this.stack.push(val);
    };
    jmpTable[Op.GE] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = left! >= right!;
      this.stack.push(val);
    };
    jmpTable[Op.LT] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = left! < right!;
      this.stack.push(val);
    };
    jmpTable[Op.LE] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = left! <= right!;
      this.stack.push(val);
    };
    jmpTable[Op.OUT] = () => {
      const value = this.stack.pop();
      Deno.stdout.writeSync(new TextEncoder().encode(`${value}`));
    };
    jmpTable[Op.IN] = () => {
      const slot = this.chunk.code[this.ip++];
      const buf = new Uint8Array(1024);
      const n = Deno.stdin.readSync(buf);
      let input = "";
      if (n !== null) {
        input = new TextDecoder().decode(buf.subarray(0, n)).trim();
      }
      this.slots[this.currBaseSlot + slot] = input;
    };
    jmpTable[Op.POP] = () => {
      this.stack.pop();
    };
    jmpTable[Op.NEG] = () => {
      const val = this.stack.pop();
      this.stack.push(-(val as number));
    };
    jmpTable[Op.INC] = () => {
      const val = this.stack.pop();
      this.stack.push((val as number) + 1);
    };
    jmpTable[Op.DEC] = () => {
      const val = this.stack.pop();
      this.stack.push((val as number) - 1);
    };
    jmpTable[Op.LOG_NOT] = () => {
      const val = this.stack.pop();
      this.stack.push(!val);
    };
    jmpTable[Op.LOG_AND] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = (left as number) & (right as number);
      this.stack.push(val);
    };
    jmpTable[Op.LOG_OR] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = (left as number) | (right as number);
      this.stack.push(val);
    };
    jmpTable[Op.LOG_XOR] = () => {
      const right = this.stack.pop();
      const left = this.stack.pop();
      const val = (left as number) ^ (right as number);
      this.stack.push(val);
    };
    jmpTable[Op.JUMP] = () => {
      const high = this.chunk.code[this.ip++];
      const low = this.chunk.code[this.ip++];
      const target = (high << 8) | low;
      this.ip = target;
    };
    jmpTable[Op.JUMP_IF_FALSE] = () => {
      const high = this.chunk.code[this.ip++];
      const low = this.chunk.code[this.ip++];
      const target = (high << 8) | low;

      const cond = this.stack.pop();
      if (!cond) this.ip = target;
    };
    jmpTable[Op.CALL] = () => {
      const high = this.chunk.code[this.ip++];
      const low = this.chunk.code[this.ip++];
      const argsCount = this.chunk.code[this.ip++];
      const target = (high << 8) | low;

      this.callStack.push({
        returnIp: this.ip,
        baseSlot: this.currBaseSlot,
      });

      const newBaseSlot = this.slots.length;

      const args = [];
      for (let i = 0; i < argsCount; i++) {
        args.push(this.stack.pop());
      }
      for (let i = argsCount - 1; i >= 0; i--) {
        this.slots.push(args[i]);
      }

      this.currBaseSlot = newBaseSlot;
      this.ip = target;
    };
    jmpTable[Op.RET] = () => {
      const value = this.stack.pop();

      if (this.callStack.length === 0) return;
      const frame = this.callStack.pop();

      const currFrameSize = this.slots.length - this.currBaseSlot;
      this.slots.splice(this.currBaseSlot, currFrameSize);

      this.ip = frame!.returnIp;
      this.currBaseSlot = frame!.baseSlot;

      if (value !== undefined) this.stack.push(value);
    };
    jmpTable[Op.EXIT] = () => {
      const status = this.stack.pop() as number;
      return Deno.exit(status);
    };
    jmpTable[Op.EVAL] = () => {
      const code = this.stack.pop() as string;
      const val = globalThis.eval(code);
      this.stack.push(val);
    };

    while (true) {
      const op = this.chunk.code[this.ip++];
      if (op === Op.HALT) return;
      jmpTable[op]();
    }
  };
}

File: src/bytecode/gvm.rs
use std::process::exit;

use crate::{bytecode::Bytecode, bytecode::Op, token::Literal};

struct CallStack {
    return_ip: usize,
    base_slot: usize,
}

pub struct GVM {
    ip: usize,
    stack: Vec<Literal>,
    slots: Vec<Literal>,
    call_stack: Vec<CallStack>,
    curr_base_slot: usize,
    bytecode: Bytecode,
}

impl GVM {
    pub fn new(bytecode: Bytecode) -> Self {
        GVM {
            ip: 0,
            stack: Vec::new(),
            slots: vec![Literal::Void; bytecode.max_slot as usize],
            call_stack: Vec::new(),
            curr_base_slot: 0,
            bytecode,
        }
    }

    pub fn run(&mut self) -> () {
        loop {
            let op = self.bytecode.chunk.code[self.ip as usize];
            self.ip += 1;
            match Op::from_u8(op).unwrap() {
                Op::LOADCONST => {
                    let idx = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    self.stack.push(self.bytecode.chunk.constants[idx].clone());
                }
                Op::LOADVAR => {
                    let slot = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    self.stack
                        .push(self.slots[self.curr_base_slot + slot].clone());
                }
                Op::STOREVAR => {
                    let slot = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    self.slots[self.curr_base_slot + slot] = self.stack.pop().unwrap();
                }
                Op::ADD => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Number(l + r));
                        }
                        (Literal::Str(l), Literal::Str(r)) => {
                            self.stack.push(Literal::Str(l + &r));
                        }
                        (Literal::Void, _) => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong types for ADD operation");
                        }
                    }
                }
                Op::SUB => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Number(l - r));
                        }
                        (Literal::Void, _) => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong types for SUB operation");
                        }
                    }
                }
                Op::MUL => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Number(l * r));
                        }
                        (Literal::Void, _) => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong types for MUL operation");
                        }
                    }
                }
                Op::DIV => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Number(l / r));
                        }
                        (Literal::Void, _) => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong types for DIV operation");
                        }
                    }
                }
                Op::EQ => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    self.stack.push(Literal::Bool(left == right));
                }
                Op::NE => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    self.stack.push(Literal::Bool(left != right));
                }
                Op::GT => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Bool(l > r));
                        }
                        _ => {
                            panic!("TypeError: Wrong types for GT operation");
                        }
                    }
                }
                Op::GE => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Bool(l >= r));
                        }
                        _ => {
                            panic!("TypeError: Wrong types for GE operation");
                        }
                    }
                }
                Op::LT => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Bool(l < r));
                        }
                        _ => {
                            panic!("TypeError: Wrong types for LT operation");
                        }
                    }
                }
                Op::LE => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack.push(Literal::Bool(l <= r));
                        }
                        _ => {
                            panic!("TypeError: Wrong types for LE operation");
                        }
                    }
                }
                Op::AND => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Bool(l), Literal::Bool(r)) => {
                            self.stack.push(Literal::Bool(l && r));
                        }
                        _ => {
                            panic!("TypeError: Wrong types for AND operation");
                        }
                    }
                }
                Op::OR => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Bool(l), Literal::Bool(r)) => {
                            self.stack.push(Literal::Bool(l || r));
                        }
                        _ => {
                            panic!("TypeError: Wrong types for OR operation");
                        }
                    }
                }
                Op::POP => {
                    self.stack.pop();
                }
                Op::NEG => {
                    let value = self.stack.pop().unwrap();
                    match value {
                        Literal::Number(v) => {
                            self.stack.push(Literal::Number(-v));
                        }
                        Literal::Void => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong type for NEG operation");
                        }
                    }
                }
                Op::POS => {}
                Op::INC => {
                    let value = self.stack.pop().unwrap();
                    match value {
                        Literal::Number(v) => {
                            self.stack.push(Literal::Number(v + 1));
                        }
                        Literal::Void => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong type for INC operation");
                        }
                    }
                }
                Op::DEC => {
                    let value = self.stack.pop().unwrap();
                    match value {
                        Literal::Number(v) => {
                            self.stack.push(Literal::Number(v - 1));
                        }
                        Literal::Void => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong type for DEC operation");
                        }
                    }
                }
                Op::LOGNOT => {
                    let value = self.stack.pop().unwrap();
                    match value {
                        Literal::Bool(v) => {
                            self.stack.push(Literal::Bool(!v));
                        }
                        Literal::Void => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong type for LOGNOT operation");
                        }
                    }
                }
                Op::LOGAND => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack
                                .push(Literal::Number((l as isize & r as isize) as i64));
                        }
                        (Literal::Bool(l), Literal::Bool(r)) => {
                            self.stack.push(Literal::Bool(l & r));
                        }
                        (Literal::Void, _) => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong types for LOGAND operation");
                        }
                    }
                }
                Op::LOGOR => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack
                                .push(Literal::Number((l as isize | r as isize) as i64));
                        }
                        (Literal::Bool(l), Literal::Bool(r)) => {
                            self.stack.push(Literal::Bool(l | r));
                        }
                        (Literal::Void, _) => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong types for LOGOR operation");
                        }
                    }
                }
                Op::LOGXOR => {
                    let right = self.stack.pop().unwrap();
                    let left = self.stack.pop().unwrap();
                    match (left, right) {
                        (Literal::Number(l), Literal::Number(r)) => {
                            self.stack
                                .push(Literal::Number((l as isize ^ r as isize) as i64));
                        }
                        (Literal::Bool(l), Literal::Bool(r)) => {
                            self.stack.push(Literal::Bool(l ^ r));
                        }
                        (Literal::Void, _) => {
                            self.stack.push(Literal::Void);
                        }
                        _ => {
                            panic!("TypeError: Wrong types for LOGXOR operation");
                        }
                    }
                }
                Op::JUMP => {
                    let high = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    let low = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    let target = (high << 8) | low;
                    self.ip = target;
                }
                Op::JUMPIFFALSE => {
                    let high = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    let low = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    let target = (high << 8) | low;
                    let condition = self.stack.pop().unwrap();
                    match condition {
                        Literal::Bool(false) => {
                            self.ip = target;
                        }
                        Literal::Bool(true) => {}
                        Literal::Void => {}
                        _ => {
                            panic!("TypeError: Wrong type for JUMP_IF_FALSE operation");
                        }
                    }
                }
                Op::CALL => {
                    let high = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    let low = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;
                    let args_count = self.bytecode.chunk.code[self.ip] as usize;
                    self.ip += 1;

                    let target = (high << 8) | low;

                    self.call_stack.push(CallStack {
                        return_ip: self.ip,
                        base_slot: self.curr_base_slot,
                    });

                    let new_base_slot = self.slots.len() as usize;

                    let args: Vec<Literal> =
                        (0..args_count).map(|_| self.stack.pop().unwrap()).collect();

                    for i in 0..args_count {
                        self.slots.push(args[args_count - i - 1].clone());
                    }

                    self.curr_base_slot = new_base_slot;
                    self.ip = target;
                }
                Op::RET => {
                    let val = self.stack.pop();

                    if self.call_stack.is_empty() {
                        panic!("RuntimeError: Call stack underflow on RET");
                    }

                    let frame = self.call_stack.pop().unwrap();

                    let curr_frame_size = self.slots.len() - self.curr_base_slot;
                    self.slots
                        .drain(self.curr_base_slot..self.curr_base_slot + curr_frame_size);

                    self.ip = frame.return_ip;
                    self.curr_base_slot = frame.base_slot;

                    if let Some(val) = val {
                        self.stack.push(val);
                    }
                }
                Op::EXIT => {
                    let status = self.stack.pop().unwrap();
                    match status {
                        Literal::Number(s) => {
                            exit(s as i32);
                        }
                        _ => {
                            exit(0);
                        }
                    }
                }
                Op::HALT => {
                    return;
                }
            }
        }
    }
}
